# DGAT-cancer
DGAT-cancer was a method for predicting cancer driver genes. It integrates the predicted pathogenicity of the somatic mutations (together with germline variants in the healthy population) with a topological network of gene expression in tumor tissues, and considering the expression levels in tumor and paracancerous tissues while employing 24 distinct features to predict novel cancer drivers.
## Algorithm
### 1.	Using Laplacian Score to select features
We used an unsupervised method, the Laplacian Score, to select features that have high power to preserve the local geometric structure of the features space. The detailed steps for the application of Laplacian Score in feature selection were as follows:  
(1) A network ![](https://latex.codecogs.com/svg.image?\mathit{N}) was constructed to connect all m candidate genes (![](https://latex.codecogs.com/svg.image?\mathit{N}\in&space;\mathbb{R}^{\mathit{m}\times&space;\mathit{m}})). For each pair of genes, we calculated the Euclidean distance, ![](https://latex.codecogs.com/svg.image?\left|\mathit{x}_{i}-\mathit{x}_{j}&space;\right|) between their feature vectors (![](https://latex.codecogs.com/svg.image?\mathit{x}_{i}) is the feature vector of gene ![](https://latex.codecogs.com/svg.image?\mathit{i}), and ![](https://latex.codecogs.com/svg.image?\mathit{x}_{j}) is the feature vector of gene ![](https://latex.codecogs.com/svg.image?\mathit{j})). Based on the Euclidean distance, if gene ![](https://latex.codecogs.com/svg.image?\mathit{i}) is among the top ![](https://latex.codecogs.com/svg.image?\mathit{k}) (here we set ![](https://latex.codecogs.com/svg.image?k=\left&space;[&space;0.01\times&space;m&space;\right&space;])) of the nearest genes to gene ![](https://latex.codecogs.com/svg.image?j), or the gene ![](https://latex.codecogs.com/svg.image?j) is among the top ![](https://latex.codecogs.com/svg.image?k) of the nearest gene to the gene ![](https://latex.codecogs.com/svg.image?i) (![](https://latex.codecogs.com/svg.image?i\neq&space;j)), we set the connection between ![](https://latex.codecogs.com/svg.image?i) and ![](https://latex.codecogs.com/svg.image?j) as ![](https://latex.codecogs.com/svg.image?N_{ij}=1). Otherwise, ![](https://latex.codecogs.com/svg.image?N_{ij}=0).  
(2) In the network ![](https://latex.codecogs.com/svg.image?N), if ![](https://latex.codecogs.com/svg.image?N_{ij}=1), we weight the edge by ![](https://latex.codecogs.com/svg.image?W_{ij}=e^{-\left\||&space;x_{i}-x_{j}\right\||^{2}) (![](https://latex.codecogs.com/svg.image?\mathit{W}\in&space;\mathbb{R}^{\mathit{m}\times&space;\mathit{m}})). Otherwise, ![](https://latex.codecogs.com/svg.image?W_{ij}=0). The weighted network reflects the local structure of the ![](https://latex.codecogs.com/svg.image?m) genes in the feature space.  
(3) Computing the Laplacian Score of each feature. Let ![](https://latex.codecogs.com/svg.image?y_{l}=\left&space;[&space;y_{l1},&space;y_{l2},&space;...,&space;y_{lm}&space;\right&space;]^{T}) denote the ![](https://latex.codecogs.com/svg.image?l)-th feature values for all m genes. We redefine ![](https://latex.codecogs.com/svg.image?y_{l}) by removing the mean from the samples as in Equation (1). The Laplacian Score (![](https://latex.codecogs.com/svg.image?LS)) is computed by Equation (2). According to the scores of each feature, we finally selected the top 20 features with the smallest ![](https://latex.codecogs.com/svg.image?LS) to be included in the prediction model for each type of cancer.  
![](https://latex.codecogs.com/svg.image?\widetilde{y}_{l}=y_{l}-\frac{y_{l}^{T}DI}{I^{T}DI}I,&space;\mathit{(1)})  
where ![](https://latex.codecogs.com/svg.image?D=diag(\sum_{j=1}^{m}W_{1j},\sum_{j=1}^{m}W_{2j},...,\sum_{j=1}^{m}W_{mj}),&space;I=\left&space;[&space;1,1,...,1&space;\right&space;]^{T}).  
![](https://latex.codecogs.com/svg.image?LS_{l}=\frac{\widetilde{y}_{l}^{T}L\widetilde{y}_{l}}{\widetilde{y}_{l}^{T}D\widetilde{y}_{l}},&space;\mathit{(2)})  
where ![](https://latex.codecogs.com/svg.image?L=D-W).  
### 2.	Data transformation  
In order to integrate multiple features of each gene into a risk score, we combined the features by Hotelling and Box-Cox transformations, which converted the feature values into p-values. For a given scaled matrix of ![](https://latex.codecogs.com/svg.image?m) genes with ![](https://latex.codecogs.com/svg.image?n) features (![](https://latex.codecogs.com/svg.image?\mathit{P}\in&space;\mathbb{R}^{\mathit{m}\times&space;\mathit{n}}), here ![](https://latex.codecogs.com/svg.image?n=20)), the Hotelling transformation is performed as:   
![](https://latex.codecogs.com/svg.image?P^{'}=U\left&space;(&space;P-IM&space;\right&space;)^{T}),  
where ![](https://latex.codecogs.com/svg.image?U=\left&space;[&space;v_{1};v_{2};...;v_{n^{'}}&space;\right&space;]^{T}) with ![](https://latex.codecogs.com/svg.image?{n}'\leqslant&space;n) as the number of chosen principal components of ![](https://latex.codecogs.com/svg.image?P) and ![](https://latex.codecogs.com/svg.image?V=\left&space;[&space;v_{1};v_{2};...;v_{n}&space;\right&space;]) are eigenvectors for the covariance matrix of ![](https://latex.codecogs.com/svg.image?P) corresponding to decreasing eigenvalues with ![](https://latex.codecogs.com/svg.image?\lambda_{1}\geq&space;&space;\lambda_{2}\geqslant...\geqslant&space;&space;\lambda_{n}). ![](https://latex.codecogs.com/svg.image?M=\left&space;[&space;\frac{1}{m}\sum_{i=1}^{m}P_{i1},\frac{1}{m}\sum_{i=1}^{m}P_{i2},...,\frac{1}{m}\sum_{i=1}^{m}P_{in}&space;\right&space;]). Thus, transformed ![](https://latex.codecogs.com/svg.image?{P}'\in&space;\mathbb{R}^{{n}'\times&space;m}). Then, Box-Cox transformation is performed as follows,  
![](https://latex.codecogs.com/svg.image?{pi}'=\left\{\begin{matrix}\frac{pi^{\beta&space;_{i}}-1}{\beta&space;_{i}},\beta&space;_{i}\neq&space;0&space;\\log\left&space;(&space;pi&space;\right&space;),\beta&space;_{i}=0\end{matrix}\right).  
where ![](https://latex.codecogs.com/svg.image?pi) is the ![](https://latex.codecogs.com/svg.image?i)-th row vector of ![](https://latex.codecogs.com/svg.image?P') and all elements of ![](https://latex.codecogs.com/svg.image?pi) vector are forced to be positive before being transformed, and ![](https://latex.codecogs.com/svg.image?\beta&space;_{i}) is the parameter for transforming ![](https://latex.codecogs.com/svg.image?pi) to ![](https://latex.codecogs.com/svg.image?pi').  
Finally, we standardized each ![](https://latex.codecogs.com/svg.image?pi') and calculated p-values for elements of ![](https://latex.codecogs.com/svg.image?pi') that is in a standard Gaussian distribution. The p-values of the elements were combined as ![](https://latex.codecogs.com/svg.image?S(j)) by Fisher's method (Fisher's combined probability test). ![](https://latex.codecogs.com/svg.image?S(j)) is termed the score of gene ![](https://latex.codecogs.com/svg.image?j).  
![](https://latex.codecogs.com/svg.image?S(j)=-ln({\chi&space;_{n'}^{2}}^{-1}(-2\sum_{i=1}^{n'}log(pi_{j}^{'}))))  
### 3.	Gibbs sampling  
The scores (![](https://latex.codecogs.com/svg.image?S)) of genes were used as conditional probabilities in Gibbs sampling to obtain a convergent probability distribution of candidate genes. In the first round of sampling, Gibbs sampling was initiated by randomly selecting ![](https://latex.codecogs.com/svg.image?m') (![](https://latex.codecogs.com/svg.image?m'\leq&space;m)) genes from the candidate genes. The ![](https://latex.codecogs.com/svg.image?m') genes were assumed to have equal probabilities of being selected. Then, in the second round of sampling, another set of ![](https://latex.codecogs.com/svg.image?m') genes was sampled from the remaining ![](https://latex.codecogs.com/svg.image?m-m') genes weighted by their scores (![](https://latex.codecogs.com/svg.image?S)). The following rounds were the same as the second round. In each round, the selected frequency (![](https://latex.codecogs.com/svg.image?\frac{No.\&space;of\&space;times\&space;the\&space;gene\&space;is\&space;selected}{No.\&space;of\&space;sampling})) of each gene was updated. All the selected frequencies of candidate genes in the i-th round were denoted as a vector (![](https://latex.codecogs.com/svg.image?m\times&space;1)), ![](https://latex.codecogs.com/svg.image?Freq_{i}). When the Euclidean norm of ![](https://latex.codecogs.com/svg.image?Freq_{i}-Freq_{i-1}) was smaller than ![](https://latex.codecogs.com/svg.image?E_{Gibbs}) (![](https://latex.codecogs.com/svg.image?E_{Gibbs}) was set as 0.01), the iteration was stopped. ![](https://latex.codecogs.com/svg.image?Freq_{last}) was assigned as the posterior probabilities (PP) of candidate genes. 







1.	Organize the data file as in cancer.list.Rdata. Gene symbols in row names and feature names in column names.
2.	Use laplacian_and_hotelling_box-box_transform.R. 
The data file was then put into the laplacian feature selection to find the most important features. Then the selected features would be transformed using Hotelling and Box-cox to generate weights of genes, which were used as sampled weights in Gibbs sampling.
3.	Use gibbs_sampling.R
The weighted score file generated in Step 2 was used as input to obtain a convergent probability distribution of candidate genes. The final selected frequencies were assigned as the posterior probabilities (PP) of candidate genes.
4.	Use random_permutation.R
The weighted score file generated in Step 2 was used as input to construct a null distribution of PP to obtain the likelihood of a given gene being a cancer driver gene. By counting the number of times that a random PP of a gene was larger than the real PP of the gene, an experience p-value was given. 
5.	The p-values were adjusted by Bonferroni correction and those genes with ![](https://latex.codecogs.com/svg.image?p_{adj})<0.01 were selected as being significant.

# Language
R (100%)
